### ISSUE TITLE
API that takes only a function

### ISSUE THREAD

**Issue by mavenlin** [y1xiaoc/fwdlap#2]
Currently the `lap` function is somehow more like the `jvp` function, which takes not only the function, but also a list of `primals`, `jacobians` and `laplacians`.

One may expect the laplace transformation to take only a function and return a transformed function that shares the same input space of the original function. For example, like the [jax.jacfwd](https://jax.readthedocs.io/en/latest/_autosummary/jax.jacfwd.html).

Is there occasions that we want to manually specify `primals`, `jacobians` and `laplacians`? Or is it due to technical difficulty?

**Comment by y1xiaoc** [comment]
Thanks for the issue! This is an intentional choice, as I want to expose lower level interface to the user, following what `jvp` and `jet` did. There are several benefits, for example:
- Current interface allows one to compute the function itself, its gradient and its laplacian at the same time. 
- There's also a partial eval version follows the same interface
- There are indeed cases where one may pass non-identity jacs_in, such as calculating kinetic energies for particles with different mass

And I think it would be fairly easy for user to implement their own functional transform interface. For functions with one variable and no aux output, something like the following would do:
```python
def laplacian(fun):
    def lapfun(x):
        eye = jnp.eye(x.size).reshape(-1, *x.shape)
        zero = fwdlap.Zero.from_value(flat_x) # or use jnp.zeros_like(x), should be the same
        return fwdlap.lap(fun, (x,), (eye,), (zero,))[-1]
    return lapfun
```
Please note that I did not test the above code so there might be an error. But the idea is like that.
For more than one variables, one may first ravel all the inputs, or refer to the implementation of jacfwd as well.

**Comment by mavenlin** [comment]
Good to know that, thanks.

BTW, I have a feeling that we could generalize to allow the user to nest grad and linear operations in any order, and compile them into the most efficient implementation. The tracing mechanism may not be enough to perform these optimizations, we may need to capture a higher order graph and perform graph rewriting on top.

e.g. 
1. rewrite `trace(nabla(nabla(f)))` to `lap(f)`
2. `jet` rules can be implemented by recursively applying the rewriting rule $D(f\circ g)(x)=D(f)(g(x))\circ D(g)(x)$.

Let me know if you happen to have any thoughts!

**Comment by y1xiaoc** [comment]
That would be very cool! However it also sounds very challenging and I don't have a clue on how to achieve that. It would definitely beyond the scope of this project (so I'll close this issue for now) but I'm happy to discuss possibilities and maybe collaborate if you have some more detailed ideas!
